src/target/nrf91.c

```diff
@@ -81,25 +158,208 @@ static void nrf91_add_flash(target_s *target, uint32_t addr, size_t length, size
 	flash->erased = 0xff;
 	target_add_flash(target, flash);
+
+	/* add separate UICR flash */
+	target_flash_s *flash_uicr = calloc(1, sizeof(*flash_uicr));
+	if (!flash_uicr) { /* calloc failed: heap exhaustion */
+		DEBUG_WARN("calloc: failed in %s\n", __func__);
+		return;
+	}
+
+	flash_uicr->start = 0xff8000U;
+	flash_uicr->length = 0x1000U;
+	flash_uicr->blocksize = 0x4U;
+	flash_uicr->erase = nrf91_uicr_flash_erase;
+	flash_uicr->write = nrf91_flash_write;
+	flash_uicr->erased = 0xff;
+	target_add_flash(target, flash_uicr);
+}
+
+static bool nrf91_mass_erase(target_s *target)
+{
+	adiv5_access_port_s *ap = cortex_ap(target);
+	adiv5_access_port_s ctrl_ap = {
+		.dp = ap->dp,
+		.apsel = 0x4U,
+	};
+
+	if (!nrf91_ctrl_ap_mass_erase(&ctrl_ap)) {
+		return false;
+	}
+
+	if (ap->dp->target_revision > 2) {
+		target_mem_write32(target, NRF91_NVMC_CONFIG, NRF91_NVMC_CONFIG_WEN);
+		while (target_mem_read32(target, NRF91_NVMC_READY) == 0) {
+			platform_delay(1);
+			DEBUG_INFO("Waiting for NVMC to become ready\n");
+		}
+
+		target_mem_write(target, 0, empty_app, empty_app_len);
+		target_mem_write32(target, NRF91_UICR_APPROTECT, NRF91_UICR_APPROTECT_UNPROTECT_VAL);
+		target_mem_write32(target, NRF91_UICR_SECUREAPPROTECT, NRF91_UICR_APPROTECT_UNPROTECT_VAL);
+
+		while (target_mem_read32(target, NRF91_NVMC_READY) == 0) {
+			platform_delay(1);
+			DEBUG_INFO("Waiting for NVMC to become ready\n");
+		}
+
+		target_mem_write32(target, NRF91_NVMC_CONFIG, NRF91_NVMC_CONFIG_REN);
+	}
+
+	return true;
+}
+
+static bool nrf91_exit_flash_mode(target_s *const target)
+{
+	adiv5_access_port_s *ap = cortex_ap(target);
+	/* Persist AP access if uninitialized (only needed for devices with hardenend APPROTECT) */
+	if (ap->dp->target_revision > 2) {
+		bool approtect_erased = target_mem_read32(target, NRF91_UICR_APPROTECT) == NRF91_UICR_ERASED_VAL;
+		bool secureapprotect_erased = target_mem_read32(target, NRF91_UICR_SECUREAPPROTECT) == NRF91_UICR_ERASED_VAL;
+
+		target_mem_write32(target, NRF91_NVMC_CONFIG, NRF91_NVMC_CONFIG_WEN);
+
+		while (target_mem_read32(target, NRF91_NVMC_READY) == 0) {
+			platform_delay(1);
+			DEBUG_INFO("Waiting for NVMC to become ready\n");
+		}
+
+		if (approtect_erased) {
+			target_mem_write32(target, NRF91_UICR_APPROTECT, NRF91_UICR_APPROTECT_UNPROTECT_VAL);
+		}
+		if (secureapprotect_erased) {
+			target_mem_write32(target, NRF91_UICR_SECUREAPPROTECT, NRF91_UICR_APPROTECT_UNPROTECT_VAL);
+		}
+
+		while (target_mem_read32(target, NRF91_NVMC_READY) == 0) {
+			platform_delay(1);
+			DEBUG_INFO("Waiting for NVMC to become ready\n");
+		}
+
+		target_mem_write32(target, NRF91_NVMC_CONFIG, NRF91_NVMC_CONFIG_REN);
+	}
+	return true;
 }
 
 bool nrf91_probe(target_s *target)
 {
 	adiv5_access_port_s *ap = cortex_ap(target);
 
-	if (ap->dp->version < 2U)
+	if (ap->dp->version < 2U || ap->dp->target_partno != NRF91_PARTNO)
 		return false;
 
-	switch (ap->dp->target_partno) {
-	case 0x90:
+#ifndef ENABLE_DEBUG
```

Comment:
This wants to be `#ifndef DEBUG_INFO_IS_NOOP` to correctly guard these memory read-backs.